runtime/devicescript-vm/dist/types.d.ts

```diff
@@ -11,25 +11,103 @@ export declare type DevsModule = EmscriptenModule & typeof Exts & {
     _jd_em_devs_verify(img: ptr, size: int32): int32;
     _jd_em_devs_client_deploy(img: ptr, size: int32): int32;
     sendPacket(pkt: Uint8Array): void;
+    /**
+     * Overrideable metod called when deployment is done.
+     * @param code error code, 0 is success.
+     */
     deployHandler(code: int32): void;
-    panicHandler(code: int32): void;
+    /**
+     * Overrideable method called when a panic code is raiserd.
+     * @param exitCode the panic code
+     */
+    panicHandler(exitCode: int32): void;
 };
 export declare module Exts {
+    /**
+     * Logging function
+     */
+    let log: (...data: any[]) => void;
+    /**
+     * Error logging function
+     */
+    let error: (...data: any[]) => void;
+    /**
+     * Callback to invoke when a packet needs to be handled by the virtual machine
+     * TODO: frame or packet?
+     * @param pkt a Jacdac frame
+     */
     function handlePacket(pkt: Uint8Array): void;
-    function setupNodeTcpSocketTransport(require: any, host: string, port: number): Promise<void>;
-    function setupWebsocketTransport(url: string, proto?: string): Promise<void>;
+    interface TransportResult {
+        /**
+         * Callback to close the transport
+         */
+        close: () => void;
+    }
+    /**
+     * Starts a packet transport over a TCP socket in a node.js application
+     * @param require module resolution function, requires "net" package
+     * @param host socket url host
+     * @param port socket port
+     */
+    function setupNodeTcpSocketTransport(require: (moduleName: string) => any, host: string, port: number): Promise<TransportResult>;
+    /**
+     * Starts a packet transport over a WebSocket using arraybuffer binary type.
+     * @param url socket url
+     * @param port socket port
+     */
+    function setupWebsocketTransport(url: string | URL, protocols?: string | string[]): Promise<TransportResult>;
+    /**
+     * Utility that converts a base64-encoded buffer into a Uint8Array
+     * TODO: nobody is using this?
+     * @param s
+     * @returns
+     */
     function b64ToBin(s: string): Uint8Array;
+    /**
+     * Deploys a DeviceScript bytecode to the virtual machine
+     * TODO: what is the difference with devsClientDeploy
+     * @param binary
+     * @returns error code, 0 if deployment is successful
+     */
     function devsDeploy(binary: Uint8Array): number;
     /**
      * Verifies the format and version of the bytecode
      * @param binary DeviceScript bytecode
-     * @returns error code, 0 if verificatino is successful
+     * @returns error code, 0 if verification is successful
      */
     function devsVerify(binary: Uint8Array): number;
+    /**
+     * Deploys a DeviceScript bytecode to the virtual machine
+     * @param binary
```

Comment:
This deploys to some other attached VM but to to self

Reply:
How do you specify which other vm?

Reply:
you don't; deploys to the first one it finds.

Reply:
I mean it's a hack...

Reply:
Marked as @internal